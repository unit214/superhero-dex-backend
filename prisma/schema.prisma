// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["native", "debian-openssl-1.1.x"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Token {
  id         Int     @id @default(autoincrement())
  address    String  @unique
  symbol     String  @default("n/a")
  name       String  @default("n/a")
  decimals   Int     @default(-1)
  malformed  Boolean @default(false)
  noContract Boolean @default(false)
  pairs0     Pair[]  @relation("Pair_Token0")
  pairs1     Pair[]  @relation("Pair_Token1")
  listed     Boolean @default(false)
}

model Pair {
  id                          Int                               @id @default(autoincrement())
  address                     String                            @unique
  token0                      Token                             @relation("Pair_Token0", fields: [t0], references: [id])
  t0                          Int
  token1                      Token                             @relation("Pair_Token1", fields: [t1], references: [id])
  t1                          Int
  liquidityInfo               PairLiquidityInfo?                @relation("Pair_Info")
  synchronized                Boolean
  liquidityInfoHistory        PairLiquidityInfoHistory[]
  liquidityInfoHistoryV2      PairLiquidityInfoHistoryV2[]
  liquidityInfoHistoryError   PairLiquidityInfoHistoryError[]
  liquidityInfoHistoryV2Error PairLiquidityInfoHistoryV2Error[]
}

model PairLiquidityInfo {
  id          Int    @id
  pair        Pair   @relation("Pair_Info", fields: [id], references: [id])
  totalSupply String
  reserve0    String
  reserve1    String
  height      Int
}

model PairLiquidityInfoHistory {
  id             Int      @id @default(autoincrement())
  pair           Pair     @relation(fields: [pairId], references: [id])
  pairId         Int
  totalSupply    String
  reserve0       String
  reserve1       String
  height         Int
  microBlockHash String
  microBlockTime BigInt
  updatedAt      DateTime @default(now()) @updatedAt

  @@unique(name: "pairIdMicroBlockHashUniqueIndex", [pairId, microBlockHash])
}

model PairLiquidityInfoHistoryError {
  id             Int      @id @default(autoincrement())
  pair           Pair     @relation(fields: [pairId], references: [id])
  pairId         Int
  microBlockHash String
  error          String
  timesOccurred  Int      @default(1)
  createdAt      DateTime @default(now())
  updatedAt      DateTime @default(now()) @updatedAt

  @@unique(name: "pairIdMicroBlockHashErrorUniqueIndex", [pairId, microBlockHash, error])
}

model PairLiquidityInfoHistoryV2 {
  id              Int      @id @default(autoincrement())
  pair            Pair     @relation(fields: [pairId], references: [id])
  pairId          Int
  eventType       String
  height          Int
  microBlockTime  BigInt
  logIndex        Int
  microBlockHash  String
  transactionHash String
  totalSupply     String?
  reserve0        String?
  reserve1        String?
  deltaReserve0   String?
  deltaReserve1   String?
  fiatPrice       String?
  createdAt       DateTime @default(now())
  updatedAt       DateTime @default(now()) @updatedAt

  @@unique(name: "pairIdMicroBlockHashLogIndexUniqueIndex", [pairId, microBlockHash, logIndex])
}

model PairLiquidityInfoHistoryV2Error {
  id             Int      @id @default(autoincrement())
  pair           Pair     @relation(fields: [pairId], references: [id])
  pairId         Int
  microBlockHash String
  logIndex       Int
  error          String
  timesOccurred  Int      @default(1)
  createdAt      DateTime @default(now())
  updatedAt      DateTime @default(now()) @updatedAt

  @@unique(name: "pairIdMicroBlockHashLogIndexUniqueIndex", [pairId, microBlockHash, logIndex, error])
}
