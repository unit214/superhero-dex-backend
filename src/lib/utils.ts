import { Decimal } from '@prisma/client/runtime/library';
import BigNumber from 'bignumber.js';

export const nonNullable = <T>(t: T | null | undefined, label?: string): T => {
  if (t == null) {
    throw new Error(
      `${label ? label + ': ' : null} non nullable value expected`,
    );
  }
  return t;
};

export const removeId = <ID, T extends { id: ID }>(t: T) => {
  const {
    id, // eslint-disable-line @typescript-eslint/no-unused-vars
    ...tail
  } = t;
  return tail;
};

export const pluralize = (count: number, noun: string, suffix = 's') =>
  `${count} ${noun}${count !== 1 ? suffix : ''}`;

const parseEnv = (x) => x && JSON.parse(x);
export const presentInvalidTokens = parseEnv(process.env.SHOW_INVALID_TOKENS);

export const numberToDecimal = (number: number): Decimal =>
  new Decimal(number.toString());

export const bigIntToDecimal = (bigInt: bigint): Decimal =>
  new Decimal(bigInt.toString());

export const decimalToBigInt = (decimal: Decimal): bigint =>
  BigInt(decimal.toFixed().toString());

export const calculateUsdValue = ({
  reserve,
  tokenAePrice,
  decimals,
  aeUsdPrice,
}: {
  reserve: string;
  tokenAePrice: string;
  decimals: number;
  aeUsdPrice: string;
}) => {
  if (!reserve || !tokenAePrice || !aeUsdPrice) {
    return '0';
  }

  return new BigNumber(reserve)
    .abs()
    .div(new BigNumber(10).pow(decimals))
    .multipliedBy(tokenAePrice)
    .multipliedBy(aeUsdPrice)
    .toString();
};
