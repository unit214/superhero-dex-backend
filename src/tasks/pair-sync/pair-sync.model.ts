import { Node } from '@aeternity/aepp-sdk';
import ContractWithMethods from '@aeternity/aepp-sdk/es/contract/Contract';

import { ContractAddress } from '@/clients/sdk-client.model';

export type MetaInfo = {
  name: string;
  symbol: string;
  decimals: bigint;
};

export type Aex9Methods = {
  meta_info: () => MetaInfo;
};

export type RouterMethods = {
  factory: () => ContractAddress;
};

export type FactoryMethods = {
  get_all_pairs: () => ContractAddress[];
};

export type PairMethods = {
  token0: () => ContractAddress;
  token1: () => ContractAddress;
  total_supply: () => bigint;
  get_reserves: () => {
    reserve0: bigint;
    reserve1: bigint;
  };
};

export type Context = {
  router: ContractWithMethods<RouterMethods>;
  factory: ContractWithMethods<FactoryMethods>;
  getPair: (
    address: ContractAddress,
  ) => Promise<ContractWithMethods<PairMethods>>;
  getToken: (
    address: ContractAddress,
  ) => Promise<ContractWithMethods<Aex9Methods>>;
  node: Node;
};
