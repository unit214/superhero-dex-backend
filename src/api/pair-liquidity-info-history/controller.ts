import {
  Controller,
  Get,
  ParseEnumPipe,
  ParseIntPipe,
  Query,
} from '@nestjs/common';
import { PairLiquidityInfoHistoryService } from './service';
import { ApiOperation, ApiQuery, ApiResponse } from '@nestjs/swagger';
import * as dto from '../../dto';
import { OrderQueryEnum } from '../../dto';
import { ContractAddress } from '../../lib/utils';

@Controller('history/liquidity')
export class PairLiquidityInfoHistoryController {
  constructor(
    private readonly pairLiquidityInfoHistoryService: PairLiquidityInfoHistoryService,
  ) {}

  @Get()
  @ApiOperation({
    summary: 'Retrieve all entries of the pair liquidity info history',
    deprecated: true,
  })
  @ApiQuery({
    name: 'limit',
    type: Number,
    description: 'Limit of history entries per page (default: 100, max: 100)',
    required: false,
  })
  @ApiQuery({
    name: 'offset',
    type: Number,
    description: 'Offset of page (default: 0)',
    required: false,
  })
  @ApiQuery({
    name: 'order',
    enum: OrderQueryEnum,
    description:
      'Sorts history entries in ascending or descending order (default: asc)',
    required: false,
  })
  @ApiQuery({
    name: 'pair-address',
    type: String,
    description: 'Retrieve only history entries for the given pair address',
    required: false,
  })
  @ApiQuery({
    name: 'height',
    type: Number,
    description: 'Retrieve only history entries for the given height',
    required: false,
  })
  @ApiQuery({
    name: 'from-block-time',
    type: Number,
    description:
      'Retrieve only history entries that are equal or newer than the given micro block time',
    required: false,
  })
  @ApiQuery({
    name: 'to-block-time',
    type: Number,
    description:
      'Retrieve only history entries that are equal or older than the given micro block time',
    required: false,
  })
  @ApiResponse({ status: 200, type: [dto.PairLiquidityInfoHistoryEntry] })
  async findAll(
    @Query('limit', new ParseIntPipe({ optional: true })) limit: number = 100,
    @Query('offset', new ParseIntPipe({ optional: true })) offset: number = 0,
    @Query('order', new ParseEnumPipe(OrderQueryEnum, { optional: true }))
    order: OrderQueryEnum = OrderQueryEnum.asc,
    @Query('pair-address') pairAddress?: ContractAddress,
    @Query('height', new ParseIntPipe({ optional: true })) height?: number,
    @Query('from-block-time', new ParseIntPipe({ optional: true }))
    fromBlockTime?: number,
    @Query('to-block-time', new ParseIntPipe({ optional: true }))
    toBlockTime?: number,
  ): Promise<dto.PairLiquidityInfoHistoryEntry[]> {
    return this.pairLiquidityInfoHistoryService
      .getAllHistoryEntries(
        Number(limit),
        Number(offset),
        order,
        pairAddress,
        height ? Number(height) : undefined,
        fromBlockTime ? BigInt(fromBlockTime) : undefined,
        toBlockTime ? BigInt(toBlockTime) : undefined,
      )
      .then((entries) =>
        entries.map((entry) => ({
          pairAddress: entry.pair.address,
          liquidityInfo: {
            totalSupply: entry.totalSupply,
            reserve0: entry.reserve0,
            reserve1: entry.reserve1,
          },
          height: entry.height,
          microBlockHash: entry.microBlockHash,
          microBlockTime: entry.microBlockTime.toString(),
        })),
      );
  }
}
