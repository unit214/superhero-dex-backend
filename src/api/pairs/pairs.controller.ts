import { CacheInterceptor, CacheTTL } from '@nestjs/cache-manager';
import {
  Controller,
  Get,
  NotFoundException,
  Param,
  Query,
  UseInterceptors,
} from '@nestjs/common';
import { ApiOperation, ApiParam, ApiQuery, ApiResponse } from '@nestjs/swagger';
import * as prisma from '@prisma/client';

import * as dto from '@/api/api.model';
import { PairsService } from '@/api/pairs/pairs.service';
import { ContractAddress } from '@/clients/sdk-client.model';
import { removeId } from '@/lib/utils';

export const toLiquidityInfoDto = (
  liquidityInfo: prisma.PairLiquidityInfo | null | undefined,
): dto.LiquidityInfo | undefined =>
  liquidityInfo ? removeId(liquidityInfo) : undefined;

@UseInterceptors(CacheInterceptor)
@Controller('pairs')
export class PairsController {
  constructor(private readonly pairsService: PairsService) {}

  @Get()
  @ApiOperation({
    summary: 'Retrieves all the pairs',
    description: `the listed pairs do not have attached any liquidity info nor the tokens full information, \
for this purpose use the individual \`pairs/:address\` route`,
  })
  @ApiQuery({
    name: 'only-listed',
    type: Boolean,
    description:
      'Retrieves only the pairs having both tokens added in the official token list',
    required: false,
  })
  @ApiQuery({
    name: 'token',
    type: String,
    description:
      'Allows to filter the pairs by a specific token address (token0 or token1)',
    required: false,
  })
  @ApiResponse({ status: 200, type: [dto.PairWithUsd] })
  @CacheTTL(24 * 60 * 60 * 1000)
  async getAllPairs(
    @Query('only-listed') onlyListedStr?: string, // false | true
    @Query('token') token?: ContractAddress,
  ): Promise<dto.PairWithUsd[]> {
    const onlyListed = !!onlyListedStr && onlyListedStr !== 'false';
    return this.pairsService.getAllPairsWithAggregation(onlyListed, token);
  }

  @Get(':address')
  @ApiOperation({
    summary: 'Gets a specific pair',
    description:
      'The pair retrieved comes also with liquidity info and full token information',
  })
  @ApiParam({
    name: 'address',
    required: true,
    example:
      process.env.DOC_PAIR ||
      'ct_2JZNDfAQHZMfoBuh32Aijd9TR8A5SHUVBzxC6x5d4sS7o8xeqN',
    ...dto.pairAddressPropertyOptions,
  })
  @ApiResponse({ status: 200, type: dto.PairWithLiquidityAndTokens })
  @ApiResponse({ status: 404 })
  async findOne(
    @Param('address') address: string,
  ): Promise<dto.PairWithLiquidityAndTokens> {
    const pair = await this.pairsService.getPair(address);
    if (!pair) {
      throw new NotFoundException('pair not found');
    }
    const token0 = removeId(pair.token0);
    const token1 = removeId(pair.token1);
    return {
      address: pair.address,
      token0,
      token1,
      synchronized: pair.synchronized,
      liquidityInfo: toLiquidityInfoDto(pair.liquidityInfo),
    };
  }
}
