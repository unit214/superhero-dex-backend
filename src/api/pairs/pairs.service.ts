import { Injectable } from '@nestjs/common';

import { PairDbService } from '@/database/pair/pair-db.service';
import { presentInvalidTokens } from '@/lib/utils';

@Injectable()
export class PairsService {
  constructor(private readonly pairDbService: PairDbService) {}
  async getAllPairsWithAggregation(onlyListed?: boolean) {
    return this.pairDbService.getAllWithConditionAndAggregations(
      presentInvalidTokens,
      onlyListed,
    );
  }

  async getAllPairsWithLiquidityInfo(onlyListed?: boolean) {
    return this.pairDbService.getAllWithLiquidityInfo(
      presentInvalidTokens,
      onlyListed,
    );
  }

  async getCountStats() {
    return {
      all: await this.pairDbService.count(presentInvalidTokens),
      synced: await this.pairDbService.count(
        presentInvalidTokens,
        'synchronized',
      ),
      listed: await this.pairDbService.count(presentInvalidTokens, 'listed'),
    };
  }

  async getPair(address: string) {
    return this.pairDbService.getOne(address);
  }

  async getTopHeight() {
    return this.pairDbService.getTopHeight();
  }
}
