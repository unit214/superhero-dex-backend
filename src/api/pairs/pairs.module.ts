import { Module } from '@nestjs/common';
import { PairsController } from './pairs.controller';
import { PairsService } from './pairs.service';

@Module({
  imports: [],
  controllers: [PairsController],
  providers: [PairsService],
})
export class PairsModule {}
