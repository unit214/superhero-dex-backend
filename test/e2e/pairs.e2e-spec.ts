import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import { sortByAddress } from '../test-utils/utils';
import * as data from '../test-utils/context-mock-data';
import { cleanDb, listToken } from '../test-utils/db-helper';
import { PrismaService } from '../../src/database/prisma.service';
import { PairSyncService } from '../../src/tasks/pair-sync/pair-sync.service';
import { PairsController } from '../../src/api/pairs/pairs.controller';
import { PairsService } from '../../src/api/pairs/pairs.service';
import { PairDbService } from '../../src/database/pair/pair-db.service';
import { TokenDbService } from '../../src/database/token/token-db.service';
import { MdwWsClientService } from '../../src/clients/mdw-ws-client.service';
import { SdkClientService } from '../../src/clients/sdk-client.service';
import { mockContext } from '../test-utils/context-mock';

// Testing method
// before all
//   - initiate nest app
// before each
//   - clean db
//   - create a common context
//   - refreshPairs
// after all
//   - close nest app
//   - disconnect from prisma
describe('PairsController', () => {
  let app: INestApplication;
  let prismaService: PrismaService;
  let pairSyncService: PairSyncService;

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [PairsController],
      providers: [
        MdwWsClientService,
        SdkClientService,
        PairDbService,
        PairDbService,
        PairsService,
        PairSyncService,
        PrismaService,
        TokenDbService,
      ],
    }).compile();

    app = module.createNestApplication();
    await app.init();

    prismaService = module.get(PrismaService);
    pairSyncService = module.get(PairSyncService);
  });

  afterAll(async () => {
    await prismaService.$disconnect();
    await app.close();
  });

  beforeEach(async () => {
    await cleanDb(prismaService);
    pairSyncService.ctx = mockContext(data.context2);
    await pairSyncService['refreshPairs']();
  });

  describe('/pairs', () => {
    it('/pairs (GET) 200 unsynced', () => {
      return request(app.getHttpServer())
        .get('/pairs')
        .expect(200)
        .expect([
          {
            address: 'ct_p1',
            token0: 'ct_t0',
            token1: 'ct_t1',
            synchronized: false,
          },
          {
            address: 'ct_p2',
            token0: 'ct_t1',
            token1: 'ct_t3',
            synchronized: false,
          },
          {
            address: 'ct_p3',
            token0: 'ct_t0',
            token1: 'ct_t3',
            synchronized: false,
          },
        ]);
    });

    it('/pairs (GET) 200 synchronized', async () => {
      await pairSyncService['refreshPairsLiquidity']();
      const response = await request(app.getHttpServer())
        .get('/pairs')
        .expect(200);

      expect(sortByAddress(JSON.parse(response.text))).toEqual([
        {
          address: 'ct_p1',
          token0: 'ct_t0',
          token1: 'ct_t1',
          synchronized: true,
        },
        {
          address: 'ct_p2',
          token0: 'ct_t1',
          token1: 'ct_t3',
          synchronized: true,
        },
        {
          address: 'ct_p3',
          token0: 'ct_t0',
          token1: 'ct_t3',
          synchronized: true,
        },
      ]);
    });

    it('/pairs (GET) 200 with new pair', async () => {
      await pairSyncService['refreshPairsLiquidity']();
      pairSyncService.ctx = mockContext(data.context21);
      await pairSyncService['refreshPairs']();
      const response = await request(app.getHttpServer())
        .get('/pairs')
        .expect(200);

      expect(sortByAddress(JSON.parse(response.text))).toEqual([
        {
          address: 'ct_p1',
          token0: 'ct_t0',
          token1: 'ct_t1',
          synchronized: true,
        },
        {
          address: 'ct_p2',
          token0: 'ct_t1',
          token1: 'ct_t3',
          synchronized: true,
        },
        {
          address: 'ct_p3',
          token0: 'ct_t0',
          token1: 'ct_t3',
          synchronized: true,
        },
        {
          address: 'ct_p4',
          synchronized: false,
          token0: 'ct_t0',
          token1: 'ct_t4',
        },
      ]);
    });

    it('/pairs (GET) 200 only-listed=true', async () => {
      await pairSyncService['refreshPairsLiquidity']();
      pairSyncService.ctx = mockContext(data.context21);
      await pairSyncService['refreshPairs']();

      let response = await request(app.getHttpServer())
        .get('/pairs?only-listed=true')
        .expect(200);

      expect(sortByAddress(JSON.parse(response.text))).toEqual([]);
      await listToken(prismaService, 'ct_t0');
      await listToken(prismaService, 'ct_t3');
      await listToken(prismaService, 'ct_t4');

      response = await request(app.getHttpServer())
        .get('/pairs?only-listed=true')
        .expect(200);
      expect(sortByAddress(JSON.parse(response.text))).toEqual([
        {
          address: 'ct_p3',
          token0: 'ct_t0',
          token1: 'ct_t3',
          synchronized: true,
        },
        {
          address: 'ct_p4',
          token0: 'ct_t0',
          token1: 'ct_t4',
          synchronized: false,
        },
      ]);
    });

    it('/pairs/by-address/ct_p1 (GET) 200 no liquidity', async () => {
      return request(app.getHttpServer())
        .get('/pairs/by-address/ct_p1')
        .expect(200)
        .expect({
          address: 'ct_p1',
          token0: {
            address: 'ct_t0',
            symbol: 'A',
            name: 'A Token',
            decimals: 18,
            listed: false,
            malformed: false,
            noContract: false,
          },
          token1: {
            address: 'ct_t1',
            symbol: 'B',
            name: 'B Token',
            decimals: 6,
            listed: false,
            malformed: false,
            noContract: false,
          },
          synchronized: false,
        });
    });

    it('/pairs/by-address/ct_p1 (GET) 200 synchronized', async () => {
      await pairSyncService['refreshPairsLiquidity']();
      return request(app.getHttpServer())
        .get('/pairs/by-address/ct_p1')
        .expect(200)
        .expect({
          address: 'ct_p1',
          token0: {
            address: 'ct_t0',
            symbol: 'A',
            name: 'A Token',
            decimals: 18,
            listed: false,
            malformed: false,
            noContract: false,
          },
          token1: {
            address: 'ct_t1',
            symbol: 'B',
            name: 'B Token',
            decimals: 6,
            listed: false,
            malformed: false,
            noContract: false,
          },
          synchronized: true,
          liquidityInfo: {
            height: 1,
            totalSupply: '2',
            reserve0: '1',
            reserve1: '2',
          },
        });
    });

    it('/pairs/by-address/ct_p1 (GET) 200 unsynchronized with liquidity', async () => {
      await pairSyncService['refreshPairsLiquidity']();
      await pairSyncService['unsyncAllPairs']();
      return request(app.getHttpServer())
        .get('/pairs/by-address/ct_p1')
        .expect(200)
        .expect({
          address: 'ct_p1',
          token0: {
            address: 'ct_t0',
            symbol: 'A',
            name: 'A Token',
            decimals: 18,
            listed: false,
            malformed: false,
            noContract: false,
          },
          token1: {
            address: 'ct_t1',
            symbol: 'B',
            name: 'B Token',
            decimals: 6,
            listed: false,
            malformed: false,
            noContract: false,
          },
          synchronized: false,
          liquidityInfo: {
            height: 1,
            totalSupply: '2',
            reserve0: '1',
            reserve1: '2',
          },
        });
    });

    it('/pairs/by-address/ct_0000 (GET) 404 not founded pair', async () => {
      return request(app.getHttpServer())
        .get('/pairs/by-address/ct_0000')
        .expect(404)
        .expect({
          statusCode: 404,
          message: 'pair not found',
          error: 'Not Found',
        });
    });
  });

  describe('/pairs/swap-routes', () => {
    it('/pairs/swap-routes/ct_t0/ct_t5 (GET) 200 no path for unexisting token ', async () => {
      return request(app.getHttpServer())
        .get('/pairs/swap-routes/ct_t0/ct_t5')
        .expect(200)
        .expect([]);
    });

    it('/pairs/swap-routes/ct_t0/ct_t5 (GET) 200 no path for unexisting pair', async () => {
      pairSyncService.ctx = mockContext({
        ...data.context2,
        tokens: data.context2.tokens.concat({
          address: 'ct_t4',
          metaInfo: {
            name: 'D Token',
            symbol: 'D',
            decimals: 18n,
          },
        }),
      });
      await pairSyncService['refreshPairs']();

      return request(app.getHttpServer())
        .get('/pairs/swap-routes/ct_t0/ct_t5')
        .expect(200)
        .expect([]);
    });

    it('/pairs/swap-routes/ct_t0/ct_t4 (GET) 200 direct path', async () => {
      pairSyncService.ctx = mockContext(data.context21);
      await pairSyncService['refreshPairs']();

      return request(app.getHttpServer())
        .get('/pairs/swap-routes/ct_t0/ct_t4')
        .expect(200)
        .expect([
          [
            {
              address: 'ct_p4',
              token0: 'ct_t0',
              token1: 'ct_t4',
              synchronized: false,
            },
          ],
        ]);
    });

    it('/pairs/swap-routes/ct_t0/ct_t3 (GET) 200 synchronized', async () => {
      pairSyncService.ctx = mockContext(data.context21);
      await pairSyncService['refreshPairsLiquidity']();

      return request(app.getHttpServer())
        .get('/pairs/swap-routes/ct_t0/ct_t1')
        .expect(200)
        .expect([
          [
            {
              address: 'ct_p1',
              token0: 'ct_t0',
              token1: 'ct_t1',
              synchronized: true,
              liquidityInfo: {
                height: 1,
                totalSupply: '2',
                reserve0: '1',
                reserve1: '2',
              },
            },
          ],
          [
            {
              address: 'ct_p3',
              token0: 'ct_t0',
              token1: 'ct_t3',
              synchronized: true,
              liquidityInfo: {
                height: 1,
                totalSupply: '3',
                reserve0: '1',
                reserve1: '3',
              },
            },
            {
              address: 'ct_p2',
              token0: 'ct_t1',
              token1: 'ct_t3',
              synchronized: true,
              liquidityInfo: {
                height: 1,
                totalSupply: '200000',
                reserve0: '10',
                reserve1: '20000',
              },
            },
          ],
        ]);
    });

    it('/pairs/swap-routes/ct_t0/ct_t1 (GET) 200 one direct path and one indirect path', async () => {
      return request(app.getHttpServer())
        .get('/pairs/swap-routes/ct_t0/ct_t1')
        .expect(200)
        .expect([
          [
            {
              address: 'ct_p1',
              token0: 'ct_t0',
              token1: 'ct_t1',
              synchronized: false,
            },
          ],
          [
            {
              address: 'ct_p3',
              token0: 'ct_t0',
              token1: 'ct_t3',
              synchronized: false,
            },
            {
              address: 'ct_p2',
              token0: 'ct_t1',
              token1: 'ct_t3',
              synchronized: false,
            },
          ],
        ]);
    });

    it('/pairs/swap-routes/ct_t0/ct_t1?only-listed=true (GET) 200 suppress some paths', async () => {
      await listToken(prismaService, 'ct_t0');
      await listToken(prismaService, 'ct_t1');
      return request(app.getHttpServer())
        .get('/pairs/swap-routes/ct_t0/ct_t1?only-listed=true')
        .expect(200)
        .expect([
          [
            {
              address: 'ct_p1',
              token0: 'ct_t0',
              token1: 'ct_t1',
              synchronized: false,
            },
          ],
        ]);
    });

    it('/pairs/swap-routes/ct_t1/ct_t2 (GET) 200 testing reverse order of tokens', async () => {
      return request(app.getHttpServer())
        .get('/pairs/swap-routes/ct_t1/ct_t0')
        .expect(200)
        .expect([
          [
            {
              address: 'ct_p1',
              token0: 'ct_t0',
              token1: 'ct_t1',
              synchronized: false,
            },
          ],
          [
            {
              address: 'ct_p2',
              token0: 'ct_t1',
              token1: 'ct_t3',
              synchronized: false,
            },
            {
              address: 'ct_p3',
              token0: 'ct_t0',
              token1: 'ct_t3',
              synchronized: false,
            },
          ],
        ]);
    });

    it('/pairs/swap-routes/ct_t0/ct_t1 (GET) 200 one direct path and multiple indirect path', async () => {
      pairSyncService.ctx = mockContext({
        ...data.context21,
        pairs: data.context21.pairs.concat({
          address: 'ct_p5',
          reserve0: 1n,
          reserve1: 4n,
          totalSupply: 1n * 4n,
          t0: 1,
          t1: 3,
        }),
      });
      await pairSyncService['refreshPairs']();

      return request(app.getHttpServer())
        .get('/pairs/swap-routes/ct_t0/ct_t1')
        .expect(200)
        .expect([
          [
            {
              address: 'ct_p1',
              token0: 'ct_t0',
              token1: 'ct_t1',
              synchronized: false,
            },
          ],
          [
            {
              address: 'ct_p3',
              token0: 'ct_t0',
              token1: 'ct_t3',
              synchronized: false,
            },
            {
              address: 'ct_p2',
              token0: 'ct_t1',
              token1: 'ct_t3',
              synchronized: false,
            },
          ],
          [
            {
              address: 'ct_p4',
              token0: 'ct_t0',
              token1: 'ct_t4',
              synchronized: false,
            },
            {
              address: 'ct_p5',
              token0: 'ct_t1',
              token1: 'ct_t4',
              synchronized: false,
            },
          ],
        ]);
    });
  });
});
