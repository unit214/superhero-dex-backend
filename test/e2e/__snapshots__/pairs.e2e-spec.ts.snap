// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PairsController GET /pairs should return all pairs 1`] = `
[
  {
    "address": "ct_pair1",
    "synchronized": true,
    "token0": "ct_token1",
    "token1": "ct_token2",
    "transactions": 2,
    "tvlUsd": "0",
    "volumeUsdAll": "0",
    "volumeUsdDay": "0",
    "volumeUsdMonth": "0",
    "volumeUsdWeek": "0",
    "volumeUsdYear": "0",
  },
  {
    "address": "ct_pair2",
    "synchronized": true,
    "token0": "ct_token2",
    "token1": "ct_token3",
    "transactions": 1,
    "tvlUsd": "0",
    "volumeUsdAll": "0",
    "volumeUsdDay": "0",
    "volumeUsdMonth": "0",
    "volumeUsdWeek": "0",
    "volumeUsdYear": "0",
  },
  {
    "address": "ct_pair3",
    "synchronized": true,
    "token0": "ct_token2",
    "token1": "ct_token5",
    "transactions": 0,
    "tvlUsd": "0",
    "volumeUsdAll": null,
    "volumeUsdDay": null,
    "volumeUsdMonth": null,
    "volumeUsdWeek": null,
    "volumeUsdYear": null,
  },
  {
    "address": "ct_pair4",
    "synchronized": true,
    "token0": "ct_token1",
    "token1": "ct_token5",
    "transactions": 0,
    "tvlUsd": "0",
    "volumeUsdAll": null,
    "volumeUsdDay": null,
    "volumeUsdMonth": null,
    "volumeUsdWeek": null,
    "volumeUsdYear": null,
  },
]
`;

exports[`PairsController GET /pairs should return only listed pairs with only-listed=true 1`] = `
[
  {
    "address": "ct_pair1",
    "synchronized": true,
    "token0": "ct_token1",
    "token1": "ct_token2",
    "transactions": 2,
    "tvlUsd": "0",
    "volumeUsdAll": "0",
    "volumeUsdDay": "0",
    "volumeUsdMonth": "0",
    "volumeUsdWeek": "0",
    "volumeUsdYear": "0",
  },
]
`;

exports[`PairsController GET /pairs/{pair_address} should return pair if it exists 1`] = `
{
  "address": "ct_pair1",
  "synchronized": true,
  "token0": {
    "address": "ct_token1",
    "decimals": 18,
    "listed": false,
    "malformed": false,
    "name": "1",
    "noContract": false,
    "symbol": "1",
  },
  "token1": {
    "address": "ct_token2",
    "decimals": 18,
    "listed": false,
    "malformed": false,
    "name": "2",
    "noContract": false,
    "symbol": "2",
  },
}
`;
